warehouse/query-core/src/test/java/datawave/query/jexl/visitors/QueryPruningVisitorTest.java

```diff
@@ -0,0 +1,291 @@
+    }
+    
+    @Test
+    public void falseDoubleAndRewriteTest() throws ParseException {
+        String query = "FIELD1 == 'x' && _NOFIELD_ == 'y' && FIELD2 == 'y'";
+        ASTJexlScript script = JexlASTHelper.parseJexlQuery(query);
+        JexlNode reduced = QueryPruningVisitor.reduce(script);
+        
+        JexlEvaluation jexlEvaluation = new JexlEvaluation(JexlStringBuildingVisitor.buildQuery(reduced), new DefaultArithmetic());
+        boolean jexlState = jexlEvaluation.apply(new Tuple3<>(new Key(), new Document(), new DatawaveJexlContext()));
+        Assert.assertFalse(jexlState);
+        
+        Assert.assertEquals("((pruned = '" + query.replaceAll("'", "\\\\'") + "') && false)", JexlStringBuildingVisitor.buildQuery(reduced));
+    }
+    
+    @Test
+    public void unknownOrTest() throws ParseException {
+        String query = "FIELD1 == 'x' || _NOFIELD_ == 'y'";
+        ASTJexlScript script = JexlASTHelper.parseJexlQuery(query);
+        Assert.assertEquals(QueryPruningVisitor.TRUTH_STATE.UNKNOWN, QueryPruningVisitor.getState(script));
+    }
+    
+    @Test
+    public void trueOrTest() throws ParseException {
+        String query = "true || _NOFIELD_ == 'y'";
+        ASTJexlScript script = JexlASTHelper.parseJexlQuery(query);
+        Assert.assertEquals(QueryPruningVisitor.TRUTH_STATE.TRUE, QueryPruningVisitor.getState(script));
+    }
+    
+    @Test
+    public void branchedTest() throws ParseException {
+        String query = "FIELD1 == 'x' || (_NOFIELD_ == 'y' && FIELD2 == 'z')";
+        ASTJexlScript script = JexlASTHelper.parseJexlQuery(query);
+        Assert.assertEquals(QueryPruningVisitor.TRUTH_STATE.UNKNOWN, QueryPruningVisitor.getState(script));
+    }
+    
+    @Test
+    public void branchedRewriteTest() throws ParseException {
+        String subtree = "_NOFIELD_ == 'y' && FIELD2 == 'z'";
+        String query = "FIELD1 == 'x' || (" + subtree + ")";
+        
+        ASTJexlScript script = JexlASTHelper.parseJexlQuery(query);
+        JexlNode reduced = QueryPruningVisitor.reduce(script);
+        
+        Assert.assertEquals("FIELD1 == 'x' || (((pruned = '" + subtree.replaceAll("'", "\\\\'") + "') && false))",
+                        JexlStringBuildingVisitor.buildQuery(reduced));
+    }
+    
+    @Test
+    public void nestedBranchRewriteTest() throws ParseException {
+        String query = "((_NOFIELD_ == 'x' || _NOFIELD_ == 'y') && _NOFIELD_ == 'z') || FIELD2 == 'z'";
+        
+        ASTJexlScript script = JexlASTHelper.parseJexlQuery(query);
+        JexlNode reduced = QueryPruningVisitor.reduce(script);
+        
+        Assert.assertEquals("(((pruned = '(_NOFIELD_ == \\'x\\' || _NOFIELD_ == \\'y\\') && _NOFIELD_ == \\'z\\'') && false)) || FIELD2 == 'z'",
+                        JexlStringBuildingVisitor.buildQuery(reduced));
+    }
+    
+    @Test
+    public void deeplyNestedBranchRewriteTest() throws ParseException {
```

Comment:
It would be cool to write a test where we have nested pruned subtrees, if I'm thinking about this correctly, we are traversing the tree depth first here, so it would be possible to prine a deeply nested range than then end up pruning a parent range (that included the low-level pruned range) as well. Does that work? Are we doing extra work in that case?

Reply:
this is exactly that test